src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncBatchingWorkQueue.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A queue where items can be added to to be processed in batches after some delay has passed.
+    /// When processing happens, all the items added since the last processing point will be passed
+    /// along to be worked on.  Rounds of processing happen serially, only starting up after a
+    /// previous round has completed.
+    /// </summary>
+    internal class AsyncBatchingWorkQueue<TItem>
```

Comment:
i extracted this into its own type as it's something i've replicated in a few OOP prs at this point.  Might as well have a simple and common impl they can all use.